plugins/polio/api/vaccines/stock_management.py

```diff
@@ -78,11 +75,169 @@ class VaccineStockManagementReadWritePerm(GenericReadWritePerm):
     write_perm = permission.POLIO_VACCINE_STOCK_MANAGEMENT_WRITE
 
 
+class StockManagementCustomFilter(filters.BaseFilterBackend):
+    def filter_queryset(self, request, queryset, _view):
+        country_id = request.GET.get("country_id")
+        vaccine_type = request.GET.get("vaccine_type")
+
+        if country_id:
+            queryset = queryset.filter(country_id=country_id)
+        if vaccine_type:
+            queryset = queryset.filter(vaccine=vaccine_type)
+
+        current_order = request.GET.get("order")
+
+        if current_order:
+            if current_order == "country_name":
+                queryset = queryset.order_by("country__name")
+            elif current_order == "-country_name":
+                queryset = queryset.order_by("-country__name")
+            elif current_order == "vaccine_type":
+                queryset = queryset.order_by("vaccine")
+            elif current_order == "-vaccine_type":
+                queryset = queryset.order_by("-vaccine")
+
+        return queryset
+
+
 class VaccineStockManagementViewSet(ModelViewSet):
     permission_classes = [VaccineStockManagementReadWritePerm]
     serializer_class = VaccineStockSerializer
     http_method_names = ["get", "head", "options"]
 
+    model = VaccineStock
+
+    filter_backends = [SearchFilter, StockManagementCustomFilter]
+    search_fields = ["vaccine", "country__name"]
+
+    @action(detail=True, methods=["get"])
+    def usable_vials(self, request, pk=None):
+        if pk is None:
+            return Response({"error": "No VaccineStock ID provided"}, status=status.HTTP_400_BAD_REQUEST)
+
+        try:
+            vaccine_stock = self.get_queryset().get(id=pk)
+        except VaccineStock.DoesNotExist:
+            return Response({"error": "VaccineStock not found"}, status=status.HTTP_404_NOT_FOUND)
+
+        # First find the corresponding VaccineRequestForms
+        vrfs = VaccineRequestForm.objects.filter(
+            campaign__country=vaccine_stock.country, vaccine_type=vaccine_stock.vaccine
+        )
+        if not vrfs.exists():
+            return Response(
+                {"error": "No VaccineRequestForm found for the given VaccineStock Country and Vaccine Type"},
+                status=status.HTTP_404_NOT_FOUND,
+            )
+
+        # Then find the corresponding VaccineArrivalReports
+        arrival_reports = VaccineArrivalReport.objects.filter(request_form__in=vrfs)
+        if not arrival_reports.exists():
+            return Response(
+                {"error": "No VaccineArrivalReport found for the given VaccineStock Country and Vaccine Type"},
+                status=status.HTTP_404_NOT_FOUND,
```

Comment:
I don't think returning a 404 in this case is correct. 
We could have a situation where they just create the stock for a vaccine/country with the initial inventory, but they haven't received the vrf or var for the upcoming campaign yet. It's especially true for the arrival report, since they can get the VRF, encode it, then wait for the vaccines to arrive

Reply:
Fixed it